rfcs/0042-config-option.md

```diff
@@ -0,0 +1,350 @@
+
+#### Default values
+
+Ideally modules should work by just setting `enable = true`, which often requires setting some default settings. These should get specified in the `config` section of the module by assigning the values to the `settings` option directly. Depending on how default settings matter, we need to set them differently and for different reasons:
+
+| Reason | How to assign | Needs to track upstream | Examples | Note |
+| --- | --- | --- | --- | --- |
+| Program would fail otherwise | `mkDefault` | No | `bootstrap_ip = "172.22.68.74"` | Equivalent to a starter configuration |
+| Needed for the module to work, NixOS specifics | **Without** `mkDefault` | No | `logger = "systemd"` `data_dir = "/var/lib/foo"` | Requires the user to use `mkForce` for overriding this, hinting that they leave supported territory |
+| Module needs value to influence other options | `mkDefault` | Yes | `port = 456` (influences `allowedTCPPorts`) | |
+
+#### Additional options for single settings
+
+One can easily add additional options that correspond to single configuration settings. This is done by defining an option as usual, then applying it to `settings` with a `mkDefault`. This approach allows users to set the value either through the specialized option or `settings`, which also means that new options can be added without having to worry about backwards-compatibility.
+
+#### An example
+
+Putting it all together, here is an example of a NixOS module that uses such an approach:
+
+```nix
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.services.foo;
+  format = lib.formats.json;
+in {
+
+  options.services.foo = {
+    enable = lib.mkEnableOption "foo service";
+
+    settings = lib.mkOption {
+      type = format.type;
+      default = {};
+      description = ''
+        Configuration for foo, see <link xlink:href="https://example.com/docs/foo"/>
+        for supported values.
+      '';
+    };
+
+    # An additional option for a setting so we have an eval error if this is missing
+    domain = lib.mkOption {
+      type = lib.types.str;
+      description = "Domain this service operates on.";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.foo.settings = {
+      # Fails at runtime without any value set
+      log_level = lib.mkDefault "WARN";
+
+      # We use systemd's `StateDirectory`, so we require this (no mkDefault)
+      data_path = "/var/lib/foo";
+
+      # We use this to open the firewall, so we need to know about the default at eval time
+      port = lib.mkDefault 2546;
+
+      # Apply our specialized setting.
+      domain = lib.mkDefault cfg.domain;
+    };
+
+    environment.etc."foo-config".source = format.generate pkgs "foo-config.json" cfg.settings;
```

Comment:
```suggestion
    environment.etc."foo/bar-config.json".source = format.generate pkgs "bar-config.json" cfg.settings;
```
This seems to be a slightly more concise (enriched) exemplification with respect to `environment.etc`.

Reply:
The RFC is not a substitute and isn't intended as one for [the documentation](https://nixos.org/manual/nixos/stable/options.html#opt-environment.etc).

Reply:
@lheckemann That doesn't sound like a particularly legit oposing argument to a pareto-efficient & formally correct improvment on this RFC. Unless there is an understanding, that only perceivedly _substantial_ improvements are legit, which I would refuse to share.